### ISSUE TITLE
BUG: scipy.linalg.svd produces spurious results for dtype float32 when using sklearn.decomposition.PCA

### ISSUE THREAD

**Issue by jackbowyer** [scipy/scipy#20059]
### Describe your issue.

I have identified a bug when using sklearn.decomposition.PCA which is reported here:
https://github.com/scikit-learn/scikit-learn/discussions/27946

I feel that it is appropriate to report the issue here also since the `svd_solver='full'` argument calls scipy.linalg.svd to perform the SVD. The sklearn discussion has gone quiet and I feel that it is necessary to update the documentation to reflect this issue. Essentially, dtype float32 input arrays produce different (worse) results than the exact same array with dtype float64 i.e. the maximum discrepancy (`np.max(np.abs(X_diff))`) changes from one dtype to the next. I have synthetic datasets that can be used to test this  (`X_t_test_noise_float32.npy` and `X_t_test_noise_float64.npy`), but I don't know how to supply them.

### Reproducing Code Example

```python
import numpy as np
from sklearn.decomposition import PCA

X = np.load('X_t_test_noise_float32.npy')
# X = np.load('X_t_test_noise_float64.npy')

# Percentage of variance to retain in PCA
variance = 0.999

# PCA model fitting and transformations
pca_model = PCA(n_components=variance, svd_solver='auto')
X_compressed = pca_model.fit_transform(X)
X_reconstructed = pca_model.inverse_transform(X_compressed)
print('Original shape:', X.shape)
print('Compressed shape:', X_compressed.shape)
print('Compression factor:', X.shape[1]//X_compressed.shape[1])

print('X_reconstructed mean:', np.mean(X_reconstructed))
print('X_reconstructed max:', np.max(X_reconstructed))
print('X_reconstructed min:', np.min(X_reconstructed))

# Calculate root-mean-square error with respect to the original data
X_diff = X - X_reconstructed
rmse = np.sqrt(np.mean(X_diff**2))
print('RMSE:', rmse)
print('Max discrepancy:', np.max(np.abs(X_diff)))
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version and system information

```shell
1.7.3 1.21.6 sys.version_info(major=3, minor=7, micro=12, releaselevel='final', serial=0)
lapack_mkl_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack']
    library_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/include']
lapack_opt_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack', 'blas', 'cblas', 'lapack']
    library_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/include']
blas_mkl_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack']
    library_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/include']
blas_opt_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack', 'blas', 'cblas', 'lapack']
    library_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/data/users/jbowyer/conda/envs/machine_learning/include']
```

**Comment by ilayn** [comment]
Without any concrete test case I would agree with what @ogauthe wrote up. You are essentially giving up precision to gain space and up to an extent a bit of speed by using less information to represent the same number. Hence the results will be necessarily less accurate. These are the epsilon levels of 32 and 64 bit floats 

```python
np.finfo(np.float32).eps
1.1920929e-07

np.finfo(np.float64).eps
2.220446049250313e-16
```

**Comment by jackbowyer** [comment]
Why has this been closed?

I have explained that I have synthetic datasets with which to test this, I just need to know how to supply them to you. The error for the synthetic datasets is around 0.5 which is bad news for my use case, but not disastrous. However, the 'real' data I am using gives errors on the order of 200 which certainly is disastrous. Surely this cannot be expected behaviour, as you put it?

**Comment by jackbowyer** [comment]
It is disastrous to my use case. 

If you are saying that it is expected that the error between original data and the SVD output can be as much as 223.5 for float32 data and 7.7 for float64 data then that's absolutely fine I guess, my point is that shouldn't this be flagged in the documentation somewhere? Is it already flagged? It would have saved me hours of troubleshooting if I'd known from the beginning that float64 data gives the desired performance.

**Comment by ilayn** [comment]
We don't know your use case hence speculating. Better take it to scikit learn folks to assess better since they can give insight to PCA parts we don't have any idea what you have.

**Comment by jackbowyer** [comment]
I came here because my scikit learn discussion has been ignored, apologies.

**Comment by matthew-brett** [comment]
@jackbowyer - can you strip out the scikit-learn parts, and post the datasets somewhere public, to give a Numpy-only reproducible example?

**Comment by jackbowyer** [comment]
Hi @matthew-brett, thank you for acknowledging my request to provide the data. I will have a go at your suggestion, it might take me a while though.

**Comment by jackbowyer** [comment]
Hi @matthew-brett, I have uploaded my code here: https://github.com/jackbowyer/svd-tests

The datasets are too large to upload even after I compress them, is there a way around this?

**Comment by ilayn** [comment]
Can you compute the largest SVD with both methods? what are the results?

**Comment by matthew-brett** [comment]
Ah - interesting - so the datasets are > 50MB even compressed?  How about something like Figshare?

**Comment by rkern** [comment]
He meant the largest singular value, `S.max()`, which is `3495910.8` (the other important number would be `S.min()`, which is `2218.186`). FWIW, I get `Max discrepancy: 0.0002746582` for the `float32` case on my machine, which is within an order of magnitude of `S.max() / S.min() * np.finfo(np.float32).eps`, which is about what we're looking for.

**Comment by rkern** [comment]
Ah, the reconstruction error (as measured by `np.linalg.norm(X_diff)` is expected to be roughly proportional to `np.linalg.norm(X) * np.finfo(X.dtype).eps` (or, equivalently, `S.max() * np.finfo(X.dtype).eps`). These values are `(0.6090522, 0.41674504)`, respectively, on my machine. See the [LAPACK documentation](https://www.netlib.org/lapack/lug/node97.html) on the subject or section 8.6.2 "Perturbation Theory and Properties" under "Computing the SVD" of [_Matrix Computations_](https://www.google.com/books/edition/Matrix_Computations/X5YfsuCWpxMC?hl=en).

This amount of detail of numerical error analysis is generally a little out of scope for our docstrings. As you can see, there are lots of ways to measure error, and only some of them have any guarantees proven about them, and many of them are pretty weak (what on earth counts as a "slowly growing function of `m` and `n`" as per the LAPACK docs?). There are some general rules of thumb that people can keep in mind, but they apply variably.

**Comment by ilayn** [comment]
The error analysis typically involves `max(m, n)` as a factor meaning as the problem size grows the accumulated numerical error increases especially since these are `O(n**3)` algorithms. So as I mentioned, I am not seeing any spurious results so far from SVD. 

Basically, tl;dr of what I am saying is that if you halve the precision, your error does not necessarily doubles but goes more than that depending on the problem properties. Here it is even worse because we are combining both the forward error and the backward error by comparing to the original. 

This line below is also very error amplifying if the problem is a bit ill-conditioned

```
rmse = np.sqrt(np.mean(X_diff**2))
```

`np.linalg.norm(X_diff)` (normalize if need be) should be used to be on the safe side from underflows and overflows. 


The error analysis of SVD is pretty much established by now, hence the typical error sources are svd separation if clustered too tightly or losing the extremely small values underflowing. 

Unless there is a definite reason for us to suspect SVD I am more than interested but here we are speculating about SVD without any concrete evidence. Hence please continue the discussion but I'll close it as this is not actionable.

**Comment by jackbowyer** [comment]
Thank you @rkern and @ilayn for your comments.

The numpy.linalg.svd approach in the svd_tests.py script I have supplied gives `Max discrepancy: 0.00024414062` which is comparable to your result and hence the numpy approach is clearly unproblematic.

However, (and sorry to bring us back to sklearn) in my pca_tests.py script I have set `variance=0.999` to retain 99.9% variance in the reconstruction. But if you set that value to something like `variance=0.9999999999` or `variance=33` so that it performs the full SVD, you will see that for the float32 input this gives:  

Original shape: (33, 4915200)  
X mean: 273.72256  
X max: 326.03778  
X min: 193.75862  
X_reconstructed mean: 273.7224  
X_reconstructed max: 326.03778  
X_reconstructed min: 193.76524  
Max discrepancy: 4.6386414  
False  

and for the float64 input this gives:  

Original shape: (33, 4915200)  
X mean: 273.7226916305241  
X max: 326.0377738046512  
X min: 193.758627599943  
X_reconstructed mean: 273.7226916305239  
X_reconstructed max: 326.03777380465124  
X_reconstructed min: 193.75862759994462  
Max discrepancy: 2.1391599602793576e-09  
True  

I don't have the theoretical understanding of the problem that you have, so just so I'm completely clear you are saying that the difference in the `Max discrepancy` for these two cases (`4.6386414` and `2.1391599602793576e-09`) is expected? For completeness I computed `numpy.linalg.norm(X_diff)` which gives `14.807638` and `7.323627092539479e-09` respectively on my machine.

I can happily use sklearn.decomposition.PCA knowing that I can only produce optimal results when I provide input arrays of dtype float64, but I will reiterate that I would have saved many hours of troubleshooting had I known this from the start.

**Comment by rkern** [comment]
Ah, the test code you posted used `np.linalg.svd()` instead of `scipy.linalg.svd()` like `PCA` uses. When I change it to use `scipy.linalg.svd()`, I _do_ see more of a discrepancy in the `float32` branch, and that would explain the differences with PCA. It turns out that `np.linalg.svd()` will use the double-precision `DGESDD` LAPACK driver regardless of whether the dtype of the input is `float64` or not. `scipy.linalg.svd()` will use the single-precision `FGESDD` driver when given a `float32` input.

The `norm(X_diff)` that I get with `scipy.linalg.svd()` is about 2500 times the `S.max()*eps`; whether that is accounted for by the "slowly growing function of `m` and `n`" multiplier, I'm not sure. Probably some of that is also from the extra accumulated numerical error from numerically reconstructing the array from `U @ S @ Vh` in `float32` arithmetic rather than the infinite-precision arithmetic implicitly assumed for that operation by that error analysis (which is not _quite_ talking about your scenario where you _numerically_ reconstruct the input with finite precision operations). Given the large `n`, this amount of error certainly does not _surprise_ me.

In general, I'm not sure that there is anything that we could have said that would have saved you much time. The statements that we can make are fairly weak and somewhat distant from what you actually want. AFAIK, there actually is no rigorous proof constraining the numerical reconstruction error published for your exact scenario. As general background knowledge, you do need to know that `float32` will involve significantly more error than `float64`, but frequently, you will just have to try it and see to find out how much.

**Comment by jackbowyer** [comment]
@rkern thank you for taking the time to check this. The reason why I used `np.linalg.svd()` was because I was requested to above. I will press on with `float64`.

**Comment by rkern** [comment]
Yeah, when they asked for "NumPy-only", what they really meant was "scikit-learn-free". 😁

I know I sort all of my Github notifications for numpy and scipy to the same folder, so sometimes our sense of context gets a little mixed up when we send off replies quickly.

**Comment by matthew-brett** [comment]
Thanks as ever, @rkern for the useful and informative discussion.
